web/src/main/webapp/heatmap3d.ts

```diff
@@ -0,0 +1,804 @@
+        chartsAndScrollbarDiv.style.display = "flex";
+        chartsAndScrollbarDiv.style.flexWrap = "nowrap";
+        chartsAndScrollbarDiv.style.justifyContent = "flex-start";
+        chartsAndScrollbarDiv.style.alignItems = "stretch";
+        this.topLevel.appendChild(chartsAndScrollbarDiv);
+        chartsAndScrollbarDiv.appendChild(this.scrollBar.getHTMLRepresentation());
+
+    }
+
+    // combine two views according to some operation
+    combine(how: CombineOperators): void {
+        let r = SelectedObject.current.getSelected();
+        if (r == null) {
+            this.page.reportError("No view selected");
+            return;
+        }
+
+        let rr = this.createRpcRequest("zip", r.remoteObjectId);
+        let renderer = (page: FullPage, operation: ICancellable) => {
+            return new Make2DHistogram(
+                page, operation,
+                [this.currentData.xData.description, this.currentData.yData.description],
+                this.tableSchema, true);
+        };
+        rr.invoke(new ZipReceiver(this.getPage(), rr, how, renderer));
+    }
+
+    public static logScale(max: number) {
+        return max > 20;
+    }
+
+    public static scale(max: number, logScale: boolean) {
+        if (logScale) {
+            let base = (max > 10000) ? 10 : 2;
+            return d3.scaleLog().base(base);
+        } else {
+            return d3.scaleLinear();
+        }
+    }
+
+    changeScale(): void {
+        this.logScale = !this.logScale;
+        this.refresh();
+    }
+
+    // show the table corresponding to the data in the heatmap
+    showTable(): void {
+        let table = new TableView(this.remoteObjectId, this.page);
+        table.setSchema(this.tableSchema);
+
+        let order =  new RecordOrder([ {
+            columnDescription: this.currentData.xData.description,
+            isAscending: true
+        }, {
+            columnDescription: this.currentData.yData.description,
+            isAscending: true
+        }]);
+        let rr = table.createNextKRequest(order, null);
+        let page = new FullPage();
+        page.setDataView(table);
+        this.page.insertAfterMe(page);
+        rr.invoke(new TableRenderer(page, table, rr, false, order));
+    }
+
+    getHTMLRepresentation(): HTMLElement {
+        return this.topLevel;
+    }
+
+    // Generates a string that encodes a call to the SVG translate method
+    static translateString(x: number, y: number): string {
+        return "translate(" + String(x) + ", " + String(y) + ")";
+    }
+
+    public refresh(): void {
+        if (this.currentData == null)
+            return;
+        this.updateView(
+            this.currentData.data,
+            this.currentData.xData,
+            this.currentData.yData,
+            this.currentData.zData,
+            this.currentData.missingData,
+            0);
+    }
+
+    public pageDown() {
+        console.log("TODO: Implement pageDown");
```

Comment:
I am confused, I thought that I saw code for making pageDown work. Is this needed?